### ISSUE TITLE
Lots of jupyter_core warnings when importing another library

### ISSUE THREAD

**Issue by stefdoerr** [jupyter/jupyter_core#71]
Hey.
Since my last conda update I am getting lots of deprecation warnings from `jupyter_core` when using pyemma (another github project).
I actually cannot determine what causes the warnings to be generated but they are getting spammed so much that the notebook crashes my Firefox.

```python
In [1]: import pyemma                                                                                                                                                                                            
/shared/sdoerr/Software/anaconda3/lib/python3.4/site-packages/jupyter_core/application.py:106: DeprecationWarning: metadata {'config': True} was set from the constructor.  Metadata should be set using the .tag() method, e.g., Int().tag(key1='value1', key2='value2')
  help="""Generate default config file."""
/shared/sdoerr/Software/anaconda3/lib/python3.4/site-packages/jupyter_core/application.py:110: DeprecationWarning: metadata {'config': True} was set from the constructor.  Metadata should be set using the .tag() method, e.g., Int().tag(key1='value1', key2='value2')
  help="Specify a config file to load."
/shared/sdoerr/Software/anaconda3/lib/python3.4/site-packages/jupyter_core/application.py:118: DeprecationWarning: metadata {'config': True} was set from the constructor.  Metadata should be set using the .tag() method, e.g., Int().tag(key1='value1', key2='value2')
  help="""Full path of a config file.""",
```
and lots more.

Here is the issue I opened on their side.
https://github.com/markovmodel/PyEMMA/issues/683

**Comment by takluyver** [comment]
It looks like all the warnings are coming from jupyter, so there's probably nothing for pyemma to do. I think we were over-eager with issuing deprecation warnings.

**Comment by stefdoerr** [comment]
Okay, I just thought that since importing jupyter_core did not throw any messages it might be related to pyemma, but yes they all come from the core code.
If you need it I can provide you the full list of warnings

**Comment by Carreau** [comment]
still unsure why it would do that at pyemma import...

**Comment by Carreau** [comment]
Hum, PyEmme is definitively [inderectly maybe](https://github.com/markovmodel/msmtools/blob/devel/msmtools/util/annotators.py#L137-L166) doing stuff it should not.

```
Traceback (most recent call last):
  File "/Users/bussonniermatthias/dev/ipython/IPython/core/interactiveshell.py", line 2885, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-1-e542aa3daeae>", line 1, in <module>
    import pyemma
  File "/Users/bussonniermatthias/anaconda3/lib/python3.5/site-packages/pyemma/__init__.py", line 28, in <module>
    from . import msm
  File "/Users/bussonniermatthias/anaconda3/lib/python3.5/site-packages/pyemma/msm/__init__.py", line 91, in <module>
    from msmtools import analysis, estimation, generation, dtraj, flux
  File "/Users/bussonniermatthias/anaconda3/lib/python3.5/site-packages/msmtools/__init__.py", line 41, in <module>
    from . import analysis
  File "/Users/bussonniermatthias/anaconda3/lib/python3.5/site-packages/msmtools/analysis/__init__.py", line 113, in <module>
    from .api import *
  File "/Users/bussonniermatthias/anaconda3/lib/python3.5/site-packages/msmtools/analysis/api.py", line 92, in <module>
    @shortcut('is_tmatrix')
  File "/Users/bussonniermatthias/anaconda3/lib/python3.5/site-packages/msmtools/util/annotators.py", line 155, in shortcut
    frame = inspect.stack()[1][0]
  File "/Users/bussonniermatthias/anaconda3/lib/python3.5/inspect.py", line 1464, in stack
    return getouterframes(sys._getframe(1), context)
  File "/Users/bussonniermatthias/anaconda3/lib/python3.5/inspect.py", line 1441, in getouterframes
    frameinfo = (frame,) + getframeinfo(frame, context)
  File "/Users/bussonniermatthias/anaconda3/lib/python3.5/inspect.py", line 1414, in getframeinfo
    lines, lnum = findsource(frame)
  File "/Users/bussonniermatthias/anaconda3/lib/python3.5/inspect.py", line 756, in findsource
    module = getmodule(object, file)
  File "/Users/bussonniermatthias/anaconda3/lib/python3.5/inspect.py", line 709, in getmodule
    if ismodule(module) and hasattr(module, '__file__'):
  File "/Users/bussonniermatthias/dev/ipython/IPython/utils/shimmodule.py", line 90, in __getattr__
    return import_item(name)
  File "/Users/bussonniermatthias/dev/ipython/IPython/utils/importstring.py", line 31, in import_item
    module = __import__(package, fromlist=[obj])
  File "/Users/bussonniermatthias/dev/notebook/notebook/__init__.py", line 25, in <module>
    from .nbextensions import install_nbextension
  File "/Users/bussonniermatthias/dev/notebook/notebook/nbextensions.py", line 227, in <module>
    from jupyter_core.application import JupyterApp
  File "/Users/bussonniermatthias/anaconda3/lib/python3.5/site-packages/jupyter_core/application.py", line 61, in <module>
    class JupyterApp(Application):
  File "/Users/bussonniermatthias/anaconda3/lib/python3.5/site-packages/jupyter_core/application.py", line 105, in JupyterApp
    raise DeprecationWarning('Foo')
DeprecationWarning: Foo
```

It also for some reason disable `warnigns` filters:

```
In [1]: import warnings

In [2]: warnings.simplefilter('error')

In [3]: import pyemma
/Users/bussonniermatthias/anaconda3/lib/python3.5/site-packages/mdtraj/formats/hdf5.py:91: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
  setattr(wrapper, '__argspec__', inspect.getargspec(f))
/Users/bussonniermatthias/anaconda3/lib/python3.5/site-packages/mdtraj/formats/hdf5.py:91: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() instead
<... no execeptions>
```

And most of errors are not from `jupyter_core` (`mdtraj/formats/hdf5.py:91:`), 

The reason is that the hacks provided by `msmtools` are so horrible (well, they do what they have to do), that they end up importing our shims, which raise deprecation warnings.

**Comment by stefdoerr** [comment]
@franknoe Can you guys check this?

**Comment by franknoe** [comment]
@Carreau , could you be more specific which msmtools hacks are horrible / what should be avoided?

**Comment by takluyver** [comment]
I think he's referring to the stack inspection in `msmtools.util.annotators`.  Looking up the caller is generally a dubious thing to do, but it does mostly work. However, we monkeypatch something in inspect to work around a critical bug, so it's triggering some of our code, and something is doing something odd with warning filters that makes normally hidden warnings visible.

I'll try to dig into this a bit more to understand what's going on.

**Comment by Carreau** [comment]
> could you be more specific which msmtools hacks are horrible / what should be avoided

the "horrible" was partially written in  a sarcastic way, but yes I am referring to looking up the caller.

**Comment by marscher** [comment]
Thanks a lot for the explanation! I think the warnings filter got mistreated and finally fixed by mdtraj/mdtraj#1030. So we can close this one now.

**Comment by franknoe** [comment]
I agree.

Am 09/02/16 um 17:55 schrieb Martin K. Scherer:
>
> Thanks a lot for the explanation! I think the warnings filter got 
> mistreated and finally fixed by mdtraj/mdtraj#1030 
> <https://github.com/mdtraj/mdtraj/pull/1030>. So we can close this one 
> now.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/jupyter/jupyter_core/issues/71#issuecomment-181955206>.
>


-- 

----------------------------------------------
Prof. Dr. Frank Noe
Head of Computational Molecular Biology group
Freie Universitaet Berlin

Phone: (+49) (0)30 838 75354
Web:   research.franknoe.de
Mail:  Arnimallee 6, 14195 Berlin, Germany
----------------------------------------------

**Comment by takluyver** [comment]
Great, thanks!
